### PR TITLE
[BFT] `ReceiptValidator` ensures `k` receipts committing to the execution result

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/6863

### Context

As summarized in the issue currently `ReceiptValidator` is not ensuring that an execution result was actually executed, by not checking the availability of receipt we allow a malicious leader to incorporate made-up execution results. This PR fixes is by introducing a parameter `k` which controls how many different execution receipts have to be committing to the execution result

### CODE REVIEW

File: cmd/consensus/main.go [hunk 1]
```diff
@@ -158,6 +159,7 @@ func main() {
 		flags.UintVar(&chunkAlpha, "chunk-alpha", flow.DefaultChunkAssignmentAlpha, "number of verifiers that should be assigned to each chunk")
 		flags.UintVar(&requiredApprovalsForSealVerification, "required-verification-seal-approvals", flow.DefaultRequiredApprovalsForSealValidation, "minimum number of approvals that are required to verify a seal")
 		flags.UintVar(&requiredApprovalsForSealConstruction, "required-construction-seal-approvals", flow.DefaultRequiredApprovalsForSealConstruction, "minimum number of approvals that are required to construct a seal")
+		flags.UintVar(&requiredReceiptsCommittingToResult, "required-receipts-committing-to-result", flow.DefaultRequiredReceiptsCommittingToExecutionResult, "minimum number of receipts that are required to be included in the block for each execution result")
```
Comment (jordanschalm) [inline cmd/consensus/main.go:162]: I don't think we have a need to make this configurable. `k` is always 1.

File: module/validation/receipt_validator.go [hunk 2]
```diff
@@ -272,12 +277,28 @@ func (v *receiptValidator) ValidatePayload(candidate *flow.Block) error {
...
 		resultID := result.ID()
 
+		// check if there are enough execution receipts included in the payload corresponding to the execution result
+		receiptsForResult := uint(len(receiptsByResult.GetGroup(resultID)))
+		if receiptsForResult > 0 {
+			if receiptsForResult < v.requiredReceiptsIncludedForExecutionResult {
+				return engine.NewInvalidInputErrorf("execution result %v has only %d receipts, but at least %d are required",
+					resultID, receiptsForResult, v.requiredReceiptsIncludedForExecutionResult)
+			}
+		} else {
```
Comment (jordanschalm) [inline module/validation/receipt_validator.go:298]:
```suggestion
		if receiptsForResult == 0 {
```

File: cmd/consensus/main.go [hunk 3]
```diff
@@ -88,6 +88,7 @@ func main() {
 		chunkAlpha                            uint
 		requiredApprovalsForSealVerification  uint
 		requiredApprovalsForSealConstruction  uint
+		requiredReceiptsCommittingToResult    uint
```
Comment (jordanschalm) [inline cmd/consensus/main.go:91]:
```suggestion
```

File: module/validation/receipt_validator_test.go [hunk 4]
```diff
@@ -745,3 +752,154 @@ func (s *ReceiptValidationSuite) TestValidateReceiptAfterBootstrap() {
...
+		Receipts: []*flow.ExecutionReceiptMeta{},
+		Results:  []*flow.ExecutionResult{&receiptB.ExecutionResult},
+	}
+
+	err := s.receiptValidator.ValidatePayload(candidate)
+	s.Require().Error(err)
+	s.Require().True(engine.IsInvalidInputError(err))
+}
+
+// TestValidateReceiptResultHasNotEnoughReceipts tests a case when a malicious leader incorporates an execution result
```
Comment (jordanschalm) [inline module/validation/receipt_validator_test.go:797]: I think we can remove this test case. If `k` is constant, then this is equivalent to the previous test.

**Review by jordanschalm — COMMENTED** [review]
Looks great, however I think we can simplify by removing the parameter `k`. 

My interpretation of the issue is that `k` was meant to abstractly describe the set of valid payload inputs the implementation needs to accept, not to actually be a parameter we add to the implementation. I don't think there is a need for `k` to be configurable.

File: module/validation/receipt_validator.go (hunk 2 above)
Comment (AlexHentschel) [reply module/validation/receipt_validator.go:298]: agree with Jordan:
for each result that is incorporated in a block, there must be at least one receipt incorporated in the same block.

File: module/validation/receipt_validator_test.go [hunk 5]
```diff
@@ -881,22 +829,9 @@ func (s *ReceiptValidationSuite) TestValidateReceiptResultHasEnoughReceipts() {
 	}
 	s.PersistedResults[result0.ID()] = result0
 
-	candidateReceipts := []*flow.ExecutionReceiptMeta{receiptB.Meta()}
```
Comment (jordanschalm) [inline module/validation/receipt_validator_test.go]: I think it is still worthwhile to test both valid cases:
- Proposer includes result and 1 receipt in block
- Proposer includes result and multiple receipts in block

File: module/validation/receipt_validator.go [hunk 6]
```diff
@@ -272,12 +274,23 @@ func (v *receiptValidator) ValidatePayload(candidate *flow.Block) error {
...
+
 	// first validate all results that were included into payload
-	// if one of results is invalid we fail the whole check because it could be violating
-	// parent-children relationship
+	// if one of results is invalid we fail the whole check because it could be violating parent-children relationship
+	// each execution
 	for i, result := range payload.Results {
 		resultID := result.ID()
 
+		// check if there are enough execution receipts included in the payload corresponding to the execution result
```
Comment (jordanschalm) [inline module/validation/receipt_validator.go:288]:
```suggestion
		// Every included result must be accompanied by a receipt with a corresponding `ResultID`, in the same block.
		// If a result is included without a corresponding receipt, it cannot be attributed to any executor.
```

**Review by jordanschalm — APPROVED** [review]
🎸

**Review by AlexHentschel — APPROVED** [review]
great code and detailed test 🎵

**Pull request merged**